scripts/upgrade-facets.js

```diff
@@ -308,55 +294,62 @@ async function main(env, facetConfig) {
     }
   }
 
-  // Diamond cut - add and replace
-  const protocolInitializationFacet = deployedFacets.find((f) => f.name == "ProtocolInitializationFacet").contract;
-
-  const calldataList = [];
-  const addresses = [];
-
-  for (const key in facets.facetsToInit) {
-    const contract = deployedFacets.find((f) => f.name == key).contract;
-
-    addresses.push(contract.address);
-    calldataList.push(contract.interface.encodeFunctionData("initialize", facets.facetsToInit[key]));
-  }
-
-  const calldataProtocolInitialization = protocolInitializationFacet.interface.encodeFunctionData("initialize", [
-    // Version
-    ethers.utils.formatBytes32String(version),
-    // Facets to call initializer
-    addresses,
-    // Calldata to facets initializer
-    calldataList,
-    // Is upgrade
-    true,
-  ]);
-
-  let transactionResponse;
-
-  // Adding and replacing are done in one diamond cut
-  transactionResponse = await diamondCutFacet
-    .connect(adminSigner)
-    .diamondCut(
-      facetCut,
-      protocolInitializationFacet.address,
-      calldataProtocolInitialization,
-      await getFees(maxPriorityFeePerGas)
-    );
+  // Get ProtocolInitializationFacet from deployedFacets when added/replaced in this upgrade or get it from contracts if already deployed
+  let protocolInitializationFacet = await getInitializationFacet(deployedFacets, contracts);
+  const facetsToInit = deployedFacets.filter((facet) => facet.initialize) ?? [];
+  const initializeCalldata = getInitiliazeCalldata(facetsToInit, version, true, protocolInitializationFacet);
+
+  // Add ProtocolInitializationFacet cut action
+  deployedFacets = deployedFacets.map((f) => {
+    if (f.name == "ProtocolInitializationFacet") {
+      switch (protocolInitializationAction) {
+        case FacetCutAction.Add:
+          f.cutAdd = getFacetAddCut(f.contract, [initializeCalldata.slice(0, 10)]);
+        case FacetCutAction.Replace:
+          f.cutReplace = getFacetReplaceCut(f.contract, [initializeCalldata.slice(0, 10)]);
+      }
+    }
+    return f;
+  });
 
-  await transactionResponse.wait(confirmations);
+  let transactionResponse = await cutDiamond(
+    diamondCutFacet.address,
+    maxPriorityFeePerGas,
+    deployedFacets,
+    protocolInitializationFacet.address,
+    initializeCalldata
+  );
 
-  // Removing is done in a separate diamond cut
+  // Completed deleed facets is done in a separate diamond cut
```

Comment:
Why is this done in a separate cut? I don't see anything preventing it to be done together with add/replace. Before there was a separate cut for every facets, therefore this was separate as well.
But now it makes sense to put it together with add/repace action. Otherwise the protocol must be paused.

Change can be as simple as adding additional argument `facetCutRemove` to `cutDiamond` function and changing first argument of  `diamondCutFacet.diamondCut(` from `cut` to `[...cut, ...facetCutRemove]

Reply:
done